syne_tune/optimizer/schedulers/searchers/kde_searcher.py

```diff
@@ -0,0 +1,275 @@
+        is determined using a midpoint heuristic.
+        If None (default), this is mapped to [dict()], a single default config
+        determined by the midpoint heuristic. If [] (empty list), no initial
+        configurations are specified.
+    """
+
+    def __init__(
+            self,
+            configspace: Dict,
+            metric: str,
+            mode: str = "min",
+            num_min_data_points: int = None,
+            top_n_percent: int = 15,
+            min_bandwidth: float = 1e-3,
+            num_candidates: int = 64,
+            bandwidth_factor: int = 3,
+            random_fraction: float = .33,
+            points_to_evaluate: Optional[List[Dict]] = None,
+            **kwargs
+    ):
+        super().__init__(configspace=configspace, metric=metric, points_to_evaluate=points_to_evaluate)
+        self.mode = mode
+        self.metric_name = metric
+        self.num_evaluations = 0
+        self.min_bandwidth = min_bandwidth
+        self.random_fraction = random_fraction
+        self.num_candidates = num_candidates
+        self.bandwidth_factor = bandwidth_factor
+        self.points_to_evaluate = points_to_evaluate
+        self.top_n_percent = top_n_percent
+        self.X = []
+        self.y = []
+        self.categorical_maps = {
+            k: {cat: i for i, cat in enumerate(v.categories)}
+            for k, v in configspace.items()
+            if isinstance(v, sp.Categorical)
+        }
+        self.inv_categorical_maps = {
+            hp: dict(zip(map.values(), map.keys())) for hp, map in self.categorical_maps.items()
+        }
+
+        self.good_kde = None
+        self.bad_kde = None
+
+        self.vartypes = []
+        for hname in self.configspace:
+            hp = self.configspace[hname]
+            if isinstance(hp, sp.Categorical):
+                self.vartypes.append(('u', len(hp.categories)))
+            if isinstance(hp, sp.Integer):
+                self.vartypes.append(('o', (hp.lower, hp.upper)))
+            if isinstance(hp, sp.Float):
+                self.vartypes.append(('c', 0))
+
+        self.num_min_data_points = len(self.vartypes) if num_min_data_points is None else num_min_data_points
+        assert self.num_min_data_points >= len(self.vartypes)
+
+    @staticmethod
+    def to_feature(configspace, config, categorical_maps):
+        def numerize(value, domain, categorical_map):
+            if isinstance(domain, sp.Categorical):
+                res = categorical_map[value] / len(domain)
+                return res
+            else:
+                return [(value - domain.lower) / (domain.upper - domain.lower)]
+
+        return np.hstack([
+            numerize(value=config[k], domain=v, categorical_map=categorical_maps.get(k, {}))
+            for k, v in configspace.items()
+            if isinstance(v, sp.Domain)
+        ])
+
+    @staticmethod
+    def from_feature(configspace, feature_vector, inv_categorical_maps):
```

Comment:
Why are from_feature, to_feature static?

Reply:
because they don't access any internal variables

Reply:
Yes, but you always pass configspace and inv_categorical_maps.

Reply:
yes good point, I make it non-static